### PR TITLE
FM-171: Added Unit tests and test data for Laboratory Handler

### PR DESCRIPTION
**Pull request by HelioStrike**
# FM-171: Added Unit tests and test data for Laboratory Handler

## Description of what I changed
Created test folder in api folder and LaboratoryHandlerTest.java .

## Issue I worked on

see https://issues.openmrs.org/browse/FM-171

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        EncounterType encounterType = new EncounterType("test", "test encounter type");
+        Context.getEncounterService().saveEncounterType(encounterType);
+        encounter.setEncounterType(encounterType);
+
+        Context.getAdministrationService().setGlobalProperty("fhir.encounter.encounterRoleUuid", role.getUuid());
+
+		EncounterService es = Context.getEncounterService();
+		es.saveEncounter(encounter);
+
+        encounterService.saveEncounter(encounter);
```
Comment (Ruhanga) [inline api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: Would be nice having the above lines nicely indented.

Comment (HelioStrike) [reply api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: @Ruhanga Somehow the code looked alright in my editor, but sending the whole code a tab backward and then a backward, indicated that lines were changed and seems to have fixed the indentation. Could you please have a look at it again?

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.sameInstance;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.verify;
+import static org.mockito.MockitoAnnotations.initMocks;
+import static org.powermock.api.mockito.PowerMockito.when;
+
```
Comment (haripriya999) [inline api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: I think repeated and unused imports can be removed.

Comment (HelioStrike) [reply api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: I was pretty sure that the PR needed some changes so I didn't bother with them earlier, but I guess now's a good time to clean up. Thanks for pointing it out @haripriya999 :)

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    /**
+     * Creates and returns Encounter, given the patient's Uuid
+     * 
+     * @param patient_uuid Uuid of the patient 
+     * @return Encounter object
+     */
+    private Encounter createEncounter(String patient_uuid)
+    {
```
Comment (PrathameshMutkure) [inline api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: Properly format it as `private Encounter createEncounter(String patient_uuid) {`

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        patientService = Context.getPatientService();
+        encounterService = Context.getEncounterService();
+    }
+
+    /**
+     * Creates and returns a Patient
+     * 
+     * @return PPatient object
+     */
+    private Patient createPatient()
```
Comment (PrathameshMutkure) [inline api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: Properly format it as `private Encounter createEncounter(String patient_uuid) {`

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+        Context.getAdministrationService().setGlobalProperty("fhir.encounter.encounterType.test", encounter.getEncounterType().getUuid());
+
+        diagnosticReport = laboratoryHandler.saveFHIRDiagnosticReport(diagnosticReport);
+        String encounterId = diagnosticReport.getId().substring("DiagnosticReport/".length());
+        laboratoryHandler.retireFHIRDiagnosticReport(encounterId);
+
+        assertTrue(encounterService.getEncounterByUuid(encounterId).isVoided());    
+    }
+}
```
[the remark this answers is not in the document]
Comment (HelioStrike) [reply api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: Nice catch! :)

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        pName.setGivenName("Tom");
+        pName.setMiddleName("E.");
+        pName.setFamilyName("Patient");
+        patient.addName(pName);
+        
+        PersonAddress pAddress = new PersonAddress();
+        pAddress.setAddress1("123 My street");
+        pAddress.setAddress2("Apt 402");
+        pAddress.setCityVillage("Anywhere city");
+        pAddress.setCountry("Some Country");
```
Comment (corneliouzbett) [inline api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: Extract these constants

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    @Test
+    public void saveFHIRDiagnosticReport_shouldSaveDiagnosticReport() {
+        LaboratoryHandler laboratoryHandler = new LaboratoryHandler();
+        Patient patient = createPatient();
+        Encounter encounter = createEncounter(patient.getUuid());
+        DiagnosticReport diagnosticReport = laboratoryHandler.getFHIRDiagnosticReportById(encounter.getUuid());
+
+        Reference ref = new Reference();
+        ref.setReference("Some reference");
+        ref.setDisplay("Some display");
```
Comment (corneliouzbett) [inline api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: Ditto

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+     * 
+     * @return PPatient object
+     */
+    private Patient createPatient() {
+        Patient patient = new Patient();
+        
+        PersonName pName = new PersonName();
+        pName.setGivenName("Tom");
+        pName.setMiddleName("E.");
+        pName.setFamilyName("Patient");
```
Comment (corneliouzbett) [inline api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: Extract constants

File: api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    public void updateFHIRDiagnosticReport_shouldUpdateDiagnosticReport() {
+        LaboratoryHandler laboratoryHandler = new LaboratoryHandler();
+        Patient patient = createPatient();
+        Encounter encounter = createEncounter(patient.getUuid());
+        DiagnosticReport diagnosticReport = laboratoryHandler.getFHIRDiagnosticReportById(encounter.getUuid());
+        diagnosticReport.setIssued(new Date(System.currentTimeMillis() - 60*1000));
+
+        Reference ref = new Reference();
+        ref.setReference("Some reference");
+        ref.setDisplay("Some display");
```
Comment (corneliouzbett) [inline api/src/test/java/org/openmrs/module/fhir/api/diagnosticreport/handler/LaboratoryHandlerTest.java]: Ditto

**Comment by HelioStrike** [comment]
I've been busy with things recently. I'll pick this PR up again when I have more time. If anyone would like to work on this, feel free to borrow the code.

**Pull request closed without merging**